File: January 2023/Day 01/Count even length/count-even-length.java
//{ Driver Code Starts
//Initial Template for Java

import java.util.*;
import java.lang.*;
import java.io.*;
class GFG
{
    public static void main(String[] args) throws IOException
    {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int T = Integer.parseInt(br.readLine().trim());
        while(T-->0)
        {
            int n = Integer.parseInt(br.readLine().trim());
            Solution ob = new Solution();
            int ans = ob.compute_value(n);
            System.out.println(ans);
        }
    }
}

// } Driver Code Ends


//User function Template for Java

class Solution
{
    // Time Complexity :- O(n.log(n));
    // Space Complexity :- O(1);
    public static final long mod = 1000000007;
    public int  compute_value(int n)
    {
        // code here
        int fact_n=1;
        for(int i=1;i<=n;i++) fact_n = multi(fact_n,i);
        
        int fact_nr=fact_n;
        int fact_r=1;
        int ans=0;
        for(int r=0;r<=n;r++) {
            int denominator = multi(fact_nr,fact_r);
            int val = divide(fact_n,denominator);
            ans = add(ans,multi(val,val));
            
            fact_r = multi(fact_r,r+1);
            if(n!=r) fact_nr = divide(fact_nr,n-r);
        }
        
        return ans;
    }
    
    public static long power(long a,long b) {
        if(b==0) return 1;
        long p = power(a,b/2);
        p = multi(p,p);
        if(b%2==1) p = multi(p,a);
        return p;
    }
    
    public static int add(long a,long b) {
        return (int)(((a%mod)+(b%mod))%mod);
    }
    
    public static int multi(long a,long b) {
        return (int)(((a%mod)*(b%mod))%mod);
    }
    
    // {1/b}%mod => {b^(-1)}%mod => power(b,mod-2);
    public static int divide(long a,long b) {
        return multi(a,power(b,mod-2));
    }
    
}

File: January 2023/Day 01/Count even length/count-even-length.cpp
//{ Driver Code Starts
#include<bits/stdc++.h>
using namespace std;

// } Driver Code Ends
class Solution{
	public:
	#define ll long long int
	ll mod = 1000000007;
	int compute_value(int n)
	{
	    // Code here
	    vector<int> fact(n+1,0);
	    fact[0]=1;
        for(int i=1;i<=n;i++) fact[i] = multi(fact[i-1],i);
        
        int ans=0,val=0;
        for(int r=0;r<=n;r++) {
            val = divide(fact[n],multi(fact[r],fact[n-r]));
            ans = add(ans,multi(val,val));
        }
        
        return ans;
	}
	
	int power(ll a,ll b) {
        if(b==0) return 1;
        int p = power(a,b/2);
        p = multi(p,p);
        if(b%2==1) p = multi(p,a);
        return p;
    }
    
    int add(ll a,ll b) {
        return (int)(((a%mod)+(b%mod))%mod);
    }
    
    int multi(ll a,ll b) {
        return (int)(((a%mod)*(b%mod))%mod);
    }
    
    // {1/b}%mod => {b^(-1)}%mod => power(b,mod-2);
    int divide(ll a,ll b) {
        return multi(a,power(b,mod-2));
    }
};

//{ Driver Code Starts.
int main(){
	int tc;
	cin >> tc;
	while(tc--){
		int n;
		cin >> n;
		Solution ob;
		int ans = ob.compute_value(n);
		cout << ans <<"\n";
	}
	return 0;
}
// } Driver Code Ends